meson.build

```diff
@@ -63,6 +63,11 @@ if use_utmpx.enabled()
 elif use_utmpx.disabled()
     mconfig_data.set('USE_UTMPX', '0')
 endif
+if use_initgroups.enabled()
+    mconfig_data.set('USE_INITGROUPS', '1')
+elif use_initgroups.disabled()
+    mconfig_data.set('USE_INITGROUPS', '0')
+endif
```

Comment:
I know `grp.h` has been defined by UNIX single spec and also it's available on most systems but looks good to check that `grp.h` is available or not with these simple command:
```meson
elif use_initgroups.auto() and compiler.check_header('grp.h')
    mconfig_data.set('USE_INITGROUPS', '1')
endif
```

Reply:
@mobin-2008 it is guaranteed by POSIX that grp.h exists. Since we target POSIX systems explicitly, and since we already require this header to build anyway, we don't need to check for the header.

But more importantly, even if the header is there, that doesn't mean that `initgroups` will be defined. Setting `USE_INITGROUPS` based on whether the header exists will just cause it to be set on all systems that Dinit can be built on.

@q66:
Assuming the function will be available makes me uneasy - I've previously built Dinit successfully on systems like Sortix, which doesn't have initgroups - but I agree it's almost universal, so I'm ok with expecting it to be available at least for now.

However, there is still a good reason for not initialising the groups especially on system-level services. Specifically, intiialising group list based on a particular user id involves iterating the entire set of groups to find each group which lists the user as a member. While this isn't so bad for a single-user system where the groups list is entirely local, it is entirely possible that the groups database is read over the network (via NIS). (With Glibc this is configured via /etc/nsswitch.conf). We really don't want this happening during startup for essential services before the network is properly set up.

So, I think initialising the supplementary groups should be a service option, not a default.

Reply:
IMO it should be default, considering 1) run-as services aren't that common in the first place 2) the case you are talking about is rare, usage of dinit with it is going to be even rarer (i'd be surprised if there was even one user like that), and even then I doubt that the overhead is going to be meaningful, and if visible at all, detrimental to service startup time 3) an environment without supplementary groups initialized will run into all sorts of potential subtle misbehaviors

Reply:
It's a common setup in institutional/corporate environments (and it can definitely cause noticeable slowdowns) but I accept that in current typical setups where Dinit is actually used it won't be an issue, and ignoring supplementary groups is not great. So, ok, you've convinced me. I still think there should be an option not to set the supplementary groups, but I can add that at a later point.

Reply:
well, one thing is the slowdown being noticeable, the other is whether it actually matters - most system services that run as another user are capable of dropping privileges by themselves (and they choose themselves if auxiliary groups are needed or not), and those that do not - to cause a meaningful slowdown, you would need a 1) run-as service that is necessary to start before login 2) ideally have multiple of them 3) the group initialization would have to take at least a few seconds

unless those conditions are met, the worst case is that your boot may be slowed down by up to a couple seconds in worst case (and typically even that won't happen) which is IMO a reasonable tradeoff - I've always been of the opinion that when given the choice between maximum performance and robustness, robustness should win unless it is proven that the performance is actually necessary

if the performance becomes an actual problem in a real scenario, we could always add a service option to prevent the group initialization, which would be fine because then the service would have to acknowledge "i know what i am doing"

Reply:
>  you would need a 1) run-as service that is necessary to start before login 2) ideally have multiple of them 3) the group initialization would have to take at least a few seconds

True, though (3) is definitely plausible and (2) isn't really necessary when the network is crawling. It would be a fair point that in that situation nobody's going to be able to log in anyway, though.

>  I've always been of the opinion that when given the choice between maximum performance and robustness, robustness should win unless it is proven that the performance is actually necessary

This is in essence why I agreed:

> So, ok, you've convinced me